azure/services/virtualnetworks/spec.go

```diff
@@ -19,59 +19,67 @@ package virtualnetworks
-	"sigs.k8s.io/cluster-api-provider-azure/azure"
 	"sigs.k8s.io/cluster-api-provider-azure/azure/converters"
 )
 
 // VNetSpec defines the specification for a Virtual Network.
 type VNetSpec struct {
 	ResourceGroup    string
 	Name             string
+	Namespace        string
 	CIDRs            []string
 	Location         string
 	ExtendedLocation *infrav1.ExtendedLocationSpec
 	ClusterName      string
 	AdditionalTags   infrav1.Tags
 }
 
-// ResourceName returns the name of the vnet.
-func (s *VNetSpec) ResourceName() string {
-	return s.Name
+// ResourceRef implements azure.ASOResourceSpecGetter.
+func (s *VNetSpec) ResourceRef() *asonetworkv1.VirtualNetwork {
+	return &asonetworkv1.VirtualNetwork{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      s.Name,
+			Namespace: s.Namespace,
+		},
+	}
 }
 
-// ResourceGroupName returns the name of the resource group.
-func (s *VNetSpec) ResourceGroupName() string {
-	return s.ResourceGroup
-}
+// Parameters implements azure.ASOResourceSpecGetter.
+func (s *VNetSpec) Parameters(ctx context.Context, existing *asonetworkv1.VirtualNetwork) (*asonetworkv1.VirtualNetwork, error) {
+	vnet := existing
+	if existing == nil {
+		vnet = &asonetworkv1.VirtualNetwork{
+			Spec: asonetworkv1.VirtualNetwork_Spec{
+				Tags: infrav1.Build(infrav1.BuildParams{
+					ClusterName: s.ClusterName,
+					Lifecycle:   infrav1.ResourceLifecycleOwned,
+					Name:        ptr.To(s.Name),
+					Role:        ptr.To(infrav1.CommonRole),
+					Additional:  s.AdditionalTags,
+				}),
+			},
+		}
+	}
 
-// OwnerResourceName is a no-op for vnets.
-func (s *VNetSpec) OwnerResourceName() string {
-	return ""
+	vnet.Spec.AzureName = s.Name
+	vnet.Spec.Owner = &genruntime.KnownResourceReference{
+		Name: s.ResourceGroup,
+	}
+	vnet.Spec.Location = ptr.To(s.Location)
+	vnet.Spec.ExtendedLocation = converters.ExtendedLocationToNetworkASO(s.ExtendedLocation)
+	vnet.Spec.AddressSpace = &asonetworkv1.AddressSpace{
+		AddressPrefixes: s.CIDRs,
+	}
+
+	return vnet, nil
 }
 
-// Parameters returns the parameters for the vnet.
-func (s *VNetSpec) Parameters(ctx context.Context, existing interface{}) (interface{}, error) {
-	if existing != nil {
-		// vnet already exists, nothing to update.
-		return nil, nil
-	}
-	return armnetwork.VirtualNetwork{
-		Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
-			ClusterName: s.ClusterName,
-			Lifecycle:   infrav1.ResourceLifecycleOwned,
-			Name:        ptr.To(s.Name),
-			Role:        ptr.To(infrav1.CommonRole),
-			Additional:  s.AdditionalTags,
-		})),
-		Location:         ptr.To(s.Location),
-		ExtendedLocation: converters.ExtendedLocationToNetworkSDK(s.ExtendedLocation),
-		Properties: &armnetwork.VirtualNetworkPropertiesFormat{
-			AddressSpace: &armnetwork.AddressSpace{
-				AddressPrefixes: azure.PtrSlice(&s.CIDRs),
-			},
-		},
-	}, nil
+// WasManaged implements azure.ASOResourceSpecGetter.
+func (s *VNetSpec) WasManaged(resource *asonetworkv1.VirtualNetwork) bool {
```

Comment:
Q: is there a reason this is `WasManaged` instead of `IsManaged`?

Reply:
`WasManaged` corresponds to the current `IsManaged` for SDK-driven services. It tells CAPZ whether or not this resource should be adopted if no ASO resource already exists so we can onboard existing workload clusters to ASO.